graphql_compiler/schema_generation/schema_graph.py

```diff
@@ -504,88 +504,69 @@ def _split_classes_by_kind(self, class_name_to_definition):
-                else:
+            class_fields = _get_class_fields(class_definition)
+            abstract = _is_abstract(class_definition)
+            property_name_to_descriptor = self._get_element_properties(
+                abstract, class_name, class_name_to_definition, kind_cls)
+
+            self._elements[class_name] = kind_cls(
+                class_name, abstract, property_name_to_descriptor, class_fields)
+
+    def _get_element_properties(self, abstract, class_name, class_name_to_definition, kind_cls):
+        """Return the properties of a SchemaElement from an OrientDB class definition."""
+        property_name_to_descriptor = {}
+        all_property_lists = (
+            class_name_to_definition[inherited_class_name]['properties']
+            for inherited_class_name in self._inheritance_sets[class_name]
+        )
+        links = {EDGE_DESTINATION_PROPERTY_NAME: [], EDGE_SOURCE_PROPERTY_NAME: []}
+        for property_definition in chain.from_iterable(all_property_lists):
+            property_name = property_definition['name']
+
+            # The only properties we allow to be redefined are the in/out properties
+            # of edge classes. All other properties may only be defined once
+            # in the entire inheritance hierarchy of any schema class, of any kind.
+            duplication_allowed = property_name in links and issubclass(kind_cls, EdgeType)
+
+            if not duplication_allowed and property_name in property_name_to_descriptor:
+                raise AssertionError(u'The property "{}" on class "{}" is defined '
+                                     u'more than once, this is not allowed!'
+                                     .format(property_name, class_name))
+
+            property_descriptor = self._create_descriptor_from_property_definition(
+                class_name, property_definition, class_name_to_definition)
+
+            if property_name in links:
+                links[property_name].append(property_descriptor)
+            else:
+                property_name_to_descriptor[property_name] = property_descriptor
+        for property_name in links:
+            elements = {
+                property_descriptor.qualifier
+                for property_descriptor in links[property_name]
+            }
+            # If there are multiple in/out properties, we choose to include the one that
+            # is a subclass of all the elements present in the in/out properties.
+            for property_descriptor in links[property_name]:
+                subclass_set = self._subclass_sets[property_descriptor.qualifier]
+                if len(elements.intersection(subclass_set)) == 1:
+                    current_descriptor = property_name_to_descriptor.get(property_name, None)
+                    if current_descriptor and current_descriptor != property_descriptor:
+                        raise AssertionError(u'There already exists property "{}" in addition '
```

Comment:
The sentence in the error is somewhat confusing. I know it was there already, but we should probably try to clarify it anyway. Good error messages are one of the hallmarks of a mature software project.

Reply:
I agree. I plan on moving all of this into its own function.  That will come in a future PR.